Skeleton driver for LTC2376

meenalparakh opened the pull request:
To add a basic driver skeleton for LTC2376 (single channel, 16-bit SAR ADC) along with its device tree overlay for Raspberry Pi3 and device tree binding documentation in yaml.

nunojsa commented:
Hi,

It seems there's a lot of unrelated commits in this PR. You probably forgot to rebase your patches onto master before doing the PR?

dbogdan commented:
@meenalparakh , you should rebase your branch (meenalparakh:ltc2376_1) on top of the current gsoc2020-adc (analogdevicesinc:gsoc2020-adc) one. You need to do a git fetch and a git rebase.

meenalparakh commented:
Thank you for the feedback. Now it has one relevant commit and I have fixed the indentation.

mirceacaprioru commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
```
add Signed-off-by tag to the commit. Also try to offer a description of this commit is doing.

mirceacaprioru commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,53 @@
+
+examples:
+  - |
+    spi@7e204000 {
```
you can remove the @address for the spi and just leave it:
```
spi {
}

mirceacaprioru commented on arch/arm/boot/dts/overlays/rpi-ltc2376-overlay.dts:
```diff
@@ -0,0 +1,30 @@
+      __overlay__ {
+           status = "disabled";
+           };
+        };
```
nitpick: add extra line for visibility

mirceacaprioru commented on arch/arm/configs/adi_bcm2709_defconfig:
```diff
@@ -1382,6 +1382,7 @@ CONFIG_AD9680=y
 CONFIG_CF_AXI_TDD=y
 CONFIG_AXI_PULSE_CAPTURE=y
 CONFIG_AXI_FMCADC5_SYNC=y
+CONFIG_LTC2376=m
```
we usually add drivers statically to the kernel not as modules.

mirceacaprioru commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+	if (!indio_dev) {
+		dev_err(&spi->dev, "iio allocation failed\n");
+		return -ENOMEM;
+	}
```
add extra line

mirceacaprioru commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+	indio_dev->modes = INDIO_DIRECT_MODE;
+	indio_dev->channels = ltc2376_channels;
+	indio_dev->num_channels = ARRAY_SIZE(ltc2376_channels);
+	indio_dev->info = &ltc2376_info;
```
add extra line

mirceacaprioru commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+	}
+	st = iio_priv(indio_dev);
+	st->spi = spi;
+	spi_set_drvdata(spi, indio_dev);
```
add extra line

nunojsa replied:
I would also only set the `spi_set_drvdata` if you really need it. In a lot of situations this is done without being really needed.  For instance, if you are planning to have some Power Management in this device you will probably need it. If you will only interface with IIO subsystem, odds are that you won't (you are already saving a reference to your spi device)...

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,53 @@
+
+  spi-max-frequency:
+    description:
+      See Documentation/devicetree/bindings/spi/spi-bus.txt
```
No need for description here. Moreover, you should specify the chip maximum speed like:
```
spi-max-frequency:
   maximum: 1000000
```

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,53 @@
+                #size-cells = <0x00 >;
+                vref-supply = <&vref>;
+                reg = <0x00>;
+                spi-max-frequency = < 0xf42400 >;
```
Normally this is given as an integer... Not to readable in hex :smile:

nunojsa commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+
+#include <linux/iio/iio.h>
+#include <linux/regulator/consumer.h>
+
```
nitpick: You can get some complains upstream about the header includes. The preferred way is to have them included in alphabetical order

nunojsa commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+	struct ltc2376_state *st;
+	struct iio_dev *indio_dev;
+
+	pr_debug("ltc2376: Hello! Entered ltc2376 probe.\n");
```
use `dev_dbg()`

nunojsa commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (!indio_dev) {
+		dev_err(&spi->dev, "iio allocation failed\n");
```
No need for the log message here...

nunojsa commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+
+static const struct of_device_id ltc2376_of_match[] = {
+	{ .compatible = "adi,ltc2376" },
+	{ },
```
You can get some complains about the extra `,`

meenalparakh commented:
Changes made-
-Added description and signed off tag in the recent commit
-in drivers/iio/adc/ltc2376.c 
     -added newlines, removed "," and arranged include statements
       (the file now looks cleaner)
     -removed log message and `spi_set_drvdata` 
     -replaced pr_debug by `dev_dbg`
-in adi,ltc2376.yaml
     -removed spi-max-frequency description and @address part from spi
     -fixed the frequency 
-in adi_bcm2709_defconfig - declared as statically built driver
-in rpi-ltc2376-overlay.dts - added an extra line
Thank you for the feedback.

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,52 @@
+                compatible = "adi,ltc2376";
+                #address-cells = <0x01>;
+                #size-cells = < 0x00 >;
+                vref-supply = <&vref>;
```
Do we need `#address-cells` and `size-cells`? If so, they should be in the bindings...

meenalparakh replied:
Thank you for pointing it out. `#address-cells` and `#size-cells` are properties for the spi bus and so should not be under ltc2376. Is my reasoning correct?
I will make this change in dtoverlay and dtbinding in the next commit.

buha commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2376.yaml:
```diff
@@ -0,0 +1,50 @@
+        #address-cells = <0x01>;
+        #size-cells = <0x00>;
+
+        ltc2376@0{
```
I got this review from upstream too, rename this to a more convenient name like adc@0, it doesn't necessarily have to be ltc2376.

buha commented on arch/arm/boot/dts/overlays/rpi-ltc2376-overlay.dts:
```diff
@@ -0,0 +1,30 @@
+			ltc2376@0{
+				compatible = "adi,ltc2376";
+				reg = <0>;
+				spi-max-frequency = <16000000>;
```
I think this also needs `spi-cpha`. It seem to me from a diagram in the datasheet that the sampling is done on the falling edge of the clock signal. Double check this please.

At this point you don't really care about this but setting it now correctly avoids an extra commit later on.

buha commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,88 @@
+	struct ltc2376_state *st;
+	struct iio_dev *indio_dev;
+
+	dev_dbg(&spi->dev, "ltc2376: Hello! Entered ltc2376 probe.\n");
```
I'm not sure this is common practice. Looks like you could remove this debug message from the commit and use this kind of messages only locally, on your stash, a debug branch etc.

nunojsa replied:
Yeah, this debug message is not really adding any useful info... It's ok when developing but not in the final driver

meenalparakh replied:
Yes, will remove this line.

buha commented:
Looks good to me, but make sure to check the other commits, after that I think we can merge.

commodo commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,86 @@
+};
+module_spi_driver(ltc2376_driver);
+
+MODULE_AUTHOR("Meenal Parakh");
```
MODULE_AUTHOR usually contains an email as well;
your preference whether to leave it like this or add an email;

commodo commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,86 @@
+	{ .compatible = "adi,ltc2376" },
+	{ }
+};
+
```
this empty line is a bit extra;

commodo commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,86 @@
+/*
+ *LTC2376 Device driver_register
+ *
+ *Copyright (C) 2020 Analog Devices, Inc.
```
nitpick: a space is usually added between the start `*` and the text;
```
/*
 * LTC2376 Device driver_register
 *
 * Copyright (C) 2020 Analog Devices, Inc.
 */
```

commodo replied:
also, title is a bit weird; it should be something like: `Linear Technology LTC2376 Low Power SAR ADC`
it's a good idea to adapt the title from the datasheet

commodo commented on drivers/iio/adc/ltc2376.c:
```diff
@@ -0,0 +1,86 @@
+
+MODULE_AUTHOR("Meenal Parakh");
+MODULE_DESCRIPTION("Linear Technology LTC2376 SAR ADC driver");
+MODULE_LICENSE("GPL");
```
this is more important for upstream than for us;
but `MODULE_LICENSE("GPL");`  == `// SPDX-License-Identifier: GPL-2.0-or-later`
and `MODULE_LICENSE("GPL v2");` == `// SPDX-License-Identifier: GPL-2.0`  (or `// SPDX-License-Identifier: GPL-2.0-only`)

i don't know if it matter much whether it's GPL-v2-only or GPL-v2-or-later, but the `SPDX-License-Identifier` and `MODULE_LICENSE()` string need to match

we can leave it as-is here [for now]; but upstream may complain;
maybe they have some scripts to check this information [now];

commodo commented:
rest looks good;
we can merge this as-is [from my side]
